### ISSUE TITLE
wasm-opt error

### ISSUE THREAD

**Issue by julianlen** [WebAssembly/binaryen#4028]
Hi, I'm having problems when running `python3 check.py`. This error has been thrown also when running
`../binaryen/bin/wasm2js ./dist/zksync-crypto-bundler_bg.wasm -o $ASM`.

I think this has something to do with the optimization but not sure.

```
[ checking wasm-opt -o notation... ]

executing: binaryen/bin/wasm-opt /binaryen/test/hello_world.wat -o a.wat -S
warning: no passes specified, not doing any work
wasm-opt(89584,0x10ef03e00) malloc: *** error for object 0x7f93e77045e0: pointer being freed was not allocated
wasm-opt(89584,0x10ef03e00) malloc: *** set a breakpoint in malloc_error_break to debug
Traceback (most recent call last):
  File "/binaryen/check.py", line 400, in <module>
    sys.exit(main())
  File "binaryen/check.py", line 383, in main
    TEST_SUITES[test]()
  File "binaryen/scripts/test/wasm_opt.py", line 31, in test_wasm_opt
    support.run_command(cmd)
  File "binaryen/scripts/test/support.py", line 182, in run_command
    raise Exception(('run_command failed (%s)' % code, out + str(err or '')))
Exception: ('run_command failed (-6)', '')
```

**Comment by kripken** [comment]
Very odd... does this happen even when you run that command in the shell?
```
bin/wasm-opt test/hello_world.wat -o a.wat -S
```
I've never seen that error before. Which platform is this on? Hopefully that platform provides some debugging tools to help figure this out on your machine, like getting a stack trace. Perhaps running `valgrind` helps?

**Comment by julianlen** [comment]
I'm running it on a MacBook PRO with SO macOS Big Sur version 11.4 and when I run

`bin/wasm-opt test/hello_world.wat -o a.wat -S`

It throws:

`bin/wasm-opt test/hello_world.wat -o a.wat -S

warning: no passes specified, not doing any work
wasm-opt(97430,0x10c4fae00) malloc: *** error for object 0x7fd4f45054f0: pointer being freed was not allocated
wasm-opt(97430,0x10c4fae00) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    97430 abort      bin/wasm-opt test/hello_world.wat -o a.wat -S`

**Comment by kripken** [comment]
Does setting a breakpoint there give useful info, as the error suggests?

(I'm not familiar with mac tools, so I'm not sure how that would be done, but hopefully it's easy.)

**Comment by MaxGraey** [comment]
That's strange. I have MacBook Pro with latest BigSur (11.5.1) and `bin/wasm-opt test/hello_world.wat -o a.wat -S` works for me without any problem

**Comment by julianlen** [comment]
I'm trying to figurate out how to set the breakpoint. Meanwhile, might it be a problem with clang or gcc? 

Here my clang version:
```
clang --version
Apple clang version 12.0.5 (clang-1205.0.22.11)
Target: x86_64-apple-darwin20.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

```
gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.5 (clang-1205.0.22.11)
Target: x86_64-apple-darwin20.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by julianlen** [comment]
I'm not sure what happened, I updated Mac OS and it works.. In parallel I installed LLVM but it didn't end when it had worked, so I will close the issue. Thank you very much for the quick response.

**Comment by julianlen** [comment]
Hi! Me again, I reinstalled the project and again:

```
[100%] Built target wasm-opt
wasm2js(13371,0x1137d3e00) malloc: *** error for object 0x7f7f092ba000: pointer being freed was not allocated
wasm2js(13371,0x1137d3e00) malloc: *** set a breakpoint in malloc_error_break to debug
./build.sh: line 24: 13371 Abort trap: 6           ../binaryen/bin/wasm2js ./dist/zksync-crypto-bundler_bg.wasm -o $ASM
```

The same for

```
executing:  /Users/julian/workspace/rskj-projects/zk/zksync/sdk/binaryen/bin/wasm-opt /Users/julian/workspace/rskj-projects/zk/zksync/sdk/binaryen/test/hello_world.wat -o a.wat -S
warning: no passes specified, not doing any work
wasm-opt(13842,0x112257e00) malloc: *** error for object 0x7f8fe4805200: pointer being freed was not allocated
wasm-opt(13842,0x112257e00) malloc: *** set a breakpoint in malloc_error_break to debug
```

**Comment by julianlen** [comment]
```
(lldb) target create "bin/wasm-opt"
Current executable set to '/Users/julian/workspace/rskj-projects/zk/zksync/sdk/binaryen/bin/wasm-opt' (x86_64).
(lldb) b malloc_error_break
Breakpoint 1: where = libsystem_malloc.dylib`malloc_error_break, address = 0x00007fff2014f9ab
(lldb) run test/hello_world.wat -o a.wat -S
Process 5197 launched: '/Users/julian/workspace/rskj-projects/zk/zksync/sdk/binaryen/bin/wasm-opt' (x86_64)
warning: no passes specified, not doing any work
wasm-opt(5197,0x100920e00) malloc: *** error for object 0x107015e00: pointer being freed was not allocated
wasm-opt(5197,0x100920e00) malloc: *** set a breakpoint in malloc_error_break to debug
Process 5197 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x00007fff2034b9ab libsystem_malloc.dylib`malloc_error_break
libsystem_malloc.dylib`malloc_error_break:
->  0x7fff2034b9ab <+0>: pushq  %rbp
    0x7fff2034b9ac <+1>: movq   %rsp, %rbp
    0x7fff2034b9af <+4>: nop
    0x7fff2034b9b0 <+5>: nopl   (%rax)
Target 0: (wasm-opt) stopped.
```

**Comment by julianlen** [comment]
Solved (again),  what I did was the following:
1. Uninstalled rust `rustup self uninstall`
2. Uninstalled [xcode](https://dzone.com/articles/how-to-completely-uninstall-xcode-from-mac)
3. Uninstalled cmake `brew uninstall cmake`
4. Looked if I had another version of LLVM installed and I found that it was installed through Brew so I ran `brew uninstall llvm`.

Then, I reinstalled rust, xcode and cmake and when I ran `bin/wasm-opt test/hello_world.wat -o a.wat -S` again there were no errors.
